### PR TITLE
New music looping behaviour

### PR DESCRIPTION
**Pull request by wench**
This pull request is for changes related to #491  altering music looping so music will not continuously loop when too far from Jukebox Eggs that started looping music

### CODE REVIEW

**Comment by DominusExult** [comment]
For documentation's sake can you elaborate on what limited and auto does?
Limited uses your change I guess, stopping the loop when the jukebox egg is too far away, right?
And auto?

**Comment by wench** [comment]
> For documentation's sake can you elaborate on what limited and auto does? Limited uses your change I guess, stopping the loop when the jukebox egg is too far away, right? And auto?

Auto is the intended new be and will be the new default, 
Music will continue to loop while in the active area of the egg and when leaving the active area the repeat flag on the song is cleared so the song will will continue to play but will not repeat when it reaches the end

Limited will also loop the song while in the active area but when leaving the active area the song will immediately stop.

Config files will be interpreted with old "no" setting becomes "never", and any other old value becomes "auto"

**Comment by wench** [comment]
I am not fixing the CodeFactor issue with Egg_object::test_unhatch() as that method is mostly a copy Paste of Egg_object::is_active() with inverted logic. It seems better if the new method stays similar to the old one and I doubt anyone cares about complex methods as the old method has a worse complexity rating

**Comment by marzojr** [comment]
> I am not fixing the CodeFactor issue with Egg_object::test_unhatch() as that method is mostly a copy Paste of Egg_object::is_active() with inverted logic. It seems better if the new method stays similar to the old one and I doubt anyone cares about complex methods as the old method has a worse complexity rating

To be fair, I disagree a lot about how these tools determine a method's complexity: if you have a function consisting of a switch with enough cases that dispatches each case to a different function will result in a "complex code" warning, despite the fact that any human looking at it will trivially understand what the code does.

I would suggest ignoring such warnings unless the code really is difficult to follow.

**Comment by wench** [comment]
One thing I forgot to put in commit message is in DEBUG build with Egg displayed turned on The Active area of Jukebox_eggs will be drawn in the world as well as the Avatar's footprint, This tends to draw lines all over the screen but it is useful for testing what the eggs are doing.

File: audio/Audio.cc [hunk 1]
```diff
@@ -215,7 +215,7 @@ Audio* Audio::get_ptr() {
 Audio::Audio() {
 	assert(self == nullptr);
 
-	string s;
+	string s, newval;
```
Comment (marzojr) [inline audio/Audio.cc:218]: Why not declare newval where it is first used (line 234)?

Comment (wench) [reply audio/Audio.cc:218]: I put the declaration in first before using it

File: objs/egg.cc [hunk 2]
```diff
@@ -1111,6 +1386,12 @@ void Egg_object::hatch(
 
 	/* end hack */
 
+	// Must not try to hatch an unhatchble egg unless unhatch() was called and
+	// removedthe hatched flag
```
Comment (marzojr) [inline objs/egg.cc:1390]: `removedthe` -> `removed the`

File: objs/egg.cc [hunk 3]
```diff
@@ -1121,7 +1402,8 @@ void Egg_object::hatch(
 			// We have been deleted, so just leave.
 			return;
 		}
-		if (flags & (1 << static_cast<int>(once))) {
+		// unhatchble eggs must not be removed here
```
Comment (marzojr) [inline objs/egg.cc:1405]: unhatchable

File: objs/egg.cc [hunk 4]
```diff
@@ -858,6 +1030,109 @@ bool Egg_object::is_active(
 	}
 }
 
+// test_unhatch pretty much inverts all the criteria compared to is_active
+// thiswas added for improve jukebox_egg behaviour to give the eggs a proximity
+// near and far trigger unless other egg types are modified this code should not
+// effect anything except jukebox_eggs
+
+bool Egg_object::test_unhatch(
```
Comment (marzojr) [inline objs/egg.cc:1038]: I wonder if the logic we have in weather effects for weather eggs should not use this instead?

**Review by marzojr — APPROVED** [review]

**Pull request merged**